rules/prevent-abbreviations.js

```diff
@@ -0,0 +1,486 @@
+'use strict';
+const defaultsDeep = require('lodash.defaultsdeep');
+const toPairs = require('lodash.topairs');
+const camelCase = require('lodash.camelcase');
+const upperfirst = require('lodash.upperfirst');
+
+const getDocsUrl = require('./utils/get-docs-url');
+const avoidCapture = require('./utils/avoid-capture');
+
+const pascalCase = str => upperfirst(camelCase(str));
+const isPascalCase = str => str === pascalCase(str);
+
+const defaultReplacements = {
+	err: {
+		error: true
+	},
+	cb: {
+		callback: true
+	},
+	opts: {
+		options: true
+	},
+	str: {
+		string: true
+	},
+	obj: {
+		object: true
+	},
+	num: {
+		number: true
+	},
+	val: {
+		value: true
+	},
+	e: {
+		event: true,
+		error: true
+	},
+	evt: {
+		event: true
+	},
+	el: {
+		element: true
+	},
+	req: {
+		request: true
+	},
+	res: {
+		response: true,
+		result: true
+	},
+	btn: {
+		button: true
+	},
+	msg: {
+		message: true
+	},
+	len: {
+		length: true
+	},
+	env: {
+		environment: true
+	},
+	dev: {
+		development: true
+	},
+	prod: {
+		production: true
+	},
+	tmp: {
+		temporary: true
+	},
+	arg: {
+		argument: true
+	},
+	args: {
+		arguments: true
+	},
+	tbl: {
+		table: true
+	},
+	db: {
+		database: true
+	},
+	ctx: {
+		context: true
+	},
+	mod: {
+		module: true
+	},
+	prop: {
+		property: true
+	}
```

Comment:
@futpib We should probably add `props` here too? My concern is that it would cause a lot of error when using React, but I guess React users could disable it. Thoughts?

Reply:
It would come up in React code, but would mostly be fixable (will mostly report variable names), since you never assign directly to `this.props` and rarely have a prop named `props` (you may want that if you define a component that accepts another component as a prop, like in [redux-form](https://redux-form.com/8.1.0/docs/api/field.md/#-code-props-object-code-optional-)). But the variable name fixing would result in a IMO weird-looking (but correct) code like this:

```js
const Foo = properies => <Bar><Buz {...properies} /></Bar>;
```
```js
class Foo extends Component {
  constructor(properies) {
    super(properies);
    this.state = { /* ... */ };
  }
}
```

But, if we implement your `fooProp` idea https://github.com/sindresorhus/eslint-plugin-unicorn/pull/237#issuecomment-467748063, we will report [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) (and some other methods for older React versions). This comes back to the "part of the interface" point from https://github.com/sindresorhus/eslint-plugin-unicorn/pull/237#issuecomment-464188799

When both React and Vue use "props", how bad really is it? It's almost like a separate word now (or is a word if we consider meanings other than "properties").

But I don't really have an opinion on this. We could ask in the original issue #169 

`attr`/`attrs` also comes to mind.

Reply:
> But, if we implement your `fooProp` idea [#237 (comment)](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/237#issuecomment-467748063), we will report [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) and `defaultProps` (and some other methods for older React versions).

Yeah, that's a problem. We should probably only implement that for variables.

> When both React and Vue use "props", how bad really is it? It's almost like a separate word now (or is a word if we consider meanings other than "properties").

Couldn't React/Vue projects just disable it though? I would personally prefer using `property` and `properties` in non-React code. React/Vue code needs to adjust other ESLint rules anyway.

> `attr`/`attrs` also comes to mind.

Yes, we should have these. I don't think that conflicts with anything.

Reply:
> We should probably only implement that for variables.

I'm a little hesitant about this. If we allow abbreviated property name, but not variable name, on every destructuring you have to convert from abbreviated to full name, and generally keep both names in mind.